include/arch/arc/arch.h

```diff
@@ -78,7 +78,128 @@ extern "C" {
 
 #define _ARCH_THREAD_STACK_BUFFER(sym) ((char *)(sym + STACK_GUARD_SIZE))
 
+#ifdef CONFIG_USERSPACE
+#ifdef CONFIG_ARC_MPU
+#ifndef _ASMLANGUAGE
+#include <arch/arc/v2/mpu/arc_mpu.h>
 
+#define K_MEM_PARTITION_P_NA_U_NA	AUX_MPU_RDP_N
+#define K_MEM_PARTITION_P_RW_U_RW	(AUX_MPU_RDP_UW | AUX_MPU_RDP_UR | \
+					 AUX_MPU_RDP_KW | AUX_MPU_RDP_KR)
+#define K_MEM_PARTITION_P_RW_U_RO	(AUX_MPU_RDP_UR | \
+					 AUX_MPU_RDP_KW | AUX_MPU_RDP_KR)
+#define K_MEM_PARTITION_P_RW_U_NA	(AUX_MPU_RDP_KW | AUX_MPU_RDP_KR)
+#define K_MEM_PARTITION_P_RO_U_RO	(AUX_MPU_RDP_UR | AUX_MPU_RDP_KR)
+#define K_MEM_PARTITION_P_RO_U_NA	(AUX_MPU_RDP_KR)
+
+/* Execution-allowed attributes */
+#define K_MEM_PARTITION_P_RWX_U_RWX	(AUX_MPU_RDP_UW | AUX_MPU_RDP_UR | \
+					 AUX_MPU_RDP_KW | AUX_MPU_RDP_KR | \
+					 AUX_MPU_RDP_KE | AUX_MPU_RDP_UE)
+#define K_MEM_PARTITION_P_RWX_U_RX	(AUX_MPU_RDP_UR | \
+					 AUX_MPU_RDP_KW | AUX_MPU_RDP_KR | \
+					 AUX_MPU_RDP_KE | AUX_MPU_RDP_UE)
+#define K_MEM_PARTITION_P_RX_U_RX	(AUX_MPU_RDP_UR | \
+					 AUX_MPU_RDP_KR | \
+					 AUX_MPU_RDP_KE | AUX_MPU_RDP_UE)
+
+#define K_MEM_PARTITION_IS_WRITABLE(attr) \
+	({ \
+		int __is_writable__; \
+		attr &= (AUX_MPU_RDP_UW | AUX_MPU_RDP_KW);\
+		switch (attr) { \
+		case (AUX_MPU_RDP_UW | AUX_MPU_RDP_KW): \
+		case AUX_MPU_RDP_UW: \
+		case AUX_MPU_RDP_KW: \
+			__is_writable__ = 1; \
+			break; \
+		default: \
+			__is_writable__ = 0; \
+			break; \
+		} \
+		__is_writable__; \
+	})
+#define K_MEM_PARTITION_IS_EXECUTABLE(attr) \
+	((attr) & (AUX_MPU_RDP_KE | AUX_MPU_RDP_UE))
+
+#endif /* _ASMLANGUAGE */
+
+#if CONFIG_ARC_MPU_VER == 2
+#define _ARCH_MEM_PARTITION_ALIGN_CHECK(start, size) \
+	BUILD_ASSERT_MSG(!(((size) & ((size) - 1))) && (size) >= STACK_ALIGN \
+		 && !((u32_t)(start) & ((size) - 1)), \
+		"the size of the partition must be power of 2" \
+		" and greater than or equal to 2048." \
```

Comment:
I think the v2 MPU can be (hardware design time) configured for 32byte minimum size in recent ARC cores, instead of the 2k checked for here.

Reply:
change 2048 to the address alignment requirement of MPU